### DISCUSSION TITLE
How do I create a json object from an array of structs?

### DISCUSSION THREAD

**Discussion by m4l490n** [nlohmann/json#2266] [Q&A] [answered]
I have been reading the documentation and experimenting with some examples and I can't get what I need. I need to create a JSON object that will be an array in which each element consists of the values of a structure.

I have the following structure:

```
typedef struct
{
    uint8_t a;
    bool c;
    string d;
    float e;
} my_struct_t;
```

And I have an array of 3 these structures with different values:

`my_struct_t struct_array[] = { {1, true, "One", 1.10}, {2, false, "Two", 2.20}, {3, true, "Three", 3.30} };`

I then need to create a JSON of all this with the form:

```
{
    "array" : [
        {
            "a" : 1,
            "c": true,
            "d":"One",
            "e":1.10
        },
        {
            "a" : 2,
            "c": false,
            "d":"Two",
            "e":2.20
        },
        {
            "a" : 3,
            "c": true,
            "d":"Three",
            "e":3.30
        }
    ]
}
```
But I'm not sure exactly how to do this since there is not an example of such a case. Could someone please explain how to achieve this or point me to the section(s) where can I see this in the documentation?

Thanks! I appreciate the help!

**Comment by nlohmann** [comment, marked as answer]
For `my_struct_t`, you can define a conversion as described in https://github.com/nlohmann/json#arbitrary-types-conversions. The `array` key for the final result needs to be set explicitly. This is the result:

```cpp
#include <iostream>
#include "json.hpp"

using json = nlohmann::json;

typedef struct { uint8_t a; bool c; float e; } my_struct_t;

void to_json(json& j, const my_struct_t& ms)
{
    j = {ms.a, ms.c, ms.e};
}

int main()
{

    my_struct_t struct_array[] = {
        {1, true,  1.10},
        {2, false, 2.20},
        {3, true,  3.30}
    };

    json result = { {"array", struct_array} };
    std::cout << result << std::endl;
}
```

Output.

```json
{"array":[[1,true,1.100000023841858],[2,false,2.200000047683716],[3,true,3.299999952316284]]}
```

Note:

- I had to remove the `char*` type as C++11 does not allow initializing char arrays with string literals.
- The precision of the output of the floats is due to an internal conversion to `double`.

**Reply by m4l490n** [reply]
Thanks for helping!

I just realized as per your comment for removing the `char*`, that I made a huge mistake in my question. I have updated it because the array I need is:

```
{
    "array" : [
        {
            "a" : 1,
            "c": true,
            "d":"One",
            "e":1.10
        },
        {
            "a" : 2,
            "c": false,
            "d":"Two",
            "e":2.20
        },
        {
            "a" : 3,
            "c": true,
            "d":"Three",
            "e":3.30
        }
}
```

So instead of removing the `char*` you need to change it to `string`.

Thanks again for helping and sorry for the inconvenience.

**Reply by nlohmann** [reply]
No. `{1,true,1.100000023841858}` is invalid JSON. `{` would start an object.

**Reply by m4l490n** [reply]
@nlohmann 

You are right and sorry for the confusion, I updated my previous response and the original question. I got confused with my reading and my testing. What I need is an array like this:

```
{
    "array" : [
        {
            "a" : 1,
            "c": true,
            "d":"One",
            "e":1.10
        },
        {
            "a" : 2,
            "c": false,
            "d":"Two",
            "e":2.20
        },
        {
            "a" : 3,
            "c": true,
            "d":"Three",
            "e":3.30
        }
    ]
}
```

**Reply by nlohmann** [reply]
Then you need to adjust the `to_json` function accordingly, like `j["a"] = ms.a;`.

**Reply by m4l490n** [reply]
Awesome! Thanks!
